src/sql-server-util/src/desc.rs

```diff
@@ -0,0 +1,727 @@
+                .into_rust_if_some("ProtoSqlServerColumnDesc::decode_type")?,
+        })
+    }
+}
+
+/// Parse a raw data type from SQL Server into a Materialize [`ScalarType`].
+///
+/// Returns the [`ScalarType`] that we'll map this column to and the [`SqlServerColumnDecodeType`]
+/// that we use to decode the raw value.
+fn parse_data_type(
+    raw: &SqlServerColumnRaw,
+) -> Result<(ScalarType, SqlServerColumnDecodeType), SqlServerError> {
+    let scalar = match raw.data_type.to_lowercase().as_str() {
+        "tinyint" => (ScalarType::Int16, SqlServerColumnDecodeType::U8),
+        "smallint" => (ScalarType::Int16, SqlServerColumnDecodeType::I16),
+        "int" => (ScalarType::Int32, SqlServerColumnDecodeType::I32),
+        "bigint" => (ScalarType::Int64, SqlServerColumnDecodeType::I64),
+        "bit" => (ScalarType::Bool, SqlServerColumnDecodeType::Bool),
+        "decimal" | "numeric" => {
+            // SQL Server supports a precision in the range of [1, 38] and then
+            // the scale is 0 <= scale <= precision.
+            //
+            // Materialize numerics are floating point with a fixed precision of 39.
+            //
+            // See: <https://learn.microsoft.com/en-us/sql/t-sql/data-types/decimal-and-numeric-transact-sql?view=sql-server-ver16#arguments>
+            if raw.precision > 38 || raw.scale > raw.precision {
+                tracing::warn!(
+                    "unexpected value from SQL Server, precision of {} and scale of {}",
+                    raw.precision,
+                    raw.scale,
+                );
+            }
+            if raw.precision > 39 {
+                let reason = format!(
+                    "precision of {} is greater than our maximum of 39",
+                    raw.precision
+                );
+                return Err(SqlServerError::UnsupportedDataType {
+                    column_name: raw.name.to_string(),
+                    column_type: raw.data_type.to_string(),
+                    reason,
+                });
+            }
+
+            let raw_scale = usize::cast_from(raw.scale);
+            let max_scale = NumericMaxScale::try_from(raw_scale).map_err(|_| {
+                SqlServerError::UnsupportedDataType {
+                    column_type: raw.data_type.to_string(),
+                    column_name: raw.name.to_string(),
+                    reason: format!("scale of {} is too large", raw.scale),
+                }
+            })?;
+            let column_type = ScalarType::Numeric {
+                max_scale: Some(max_scale),
+            };
+
+            (column_type, SqlServerColumnDecodeType::Numeric)
+        }
+        "real" => (ScalarType::Float32, SqlServerColumnDecodeType::F32),
+        "double" => (ScalarType::Float64, SqlServerColumnDecodeType::F64),
+        "char" | "nchar" | "varchar" | "nvarchar" | "text" | "sysname" => {
+            // When the `max_length` is -1 SQL Server will not present us with the "before" value
```

Comment:
What is the "before" value in this context?

Reply:
Updated the comment, hopefully it makes more sense now!

Basically for sources when an `UPDATE` or `DELETE` occurs in the upstream DB we need to get the old value of the row as it existed before it was updated or deleted so we can emit `(old_val, -1)` into the data flow. In SQL Server these data types don't always include the old value, so for the moment we can't support them.